crates/holochain_persistence_lmdb/src/txn/lmdb.rs

```diff
@@ -0,0 +1,231 @@
+
+        if maybe_content.is_some() {
+            return Ok(maybe_content);
+        }
+
+        let maybe_content = self.cas_db.fetch(address)?;
+
+        if let Some(content) = maybe_content {
+            self.staging_cas_db.add(&content)?;
+            Ok(Some(content))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_id(&self) -> uuid::Uuid {
+        self.cas_db.get_id()
+    }
+}
+
+impl<A: Attribute + serde::de::DeserializeOwned> EntityAttributeValueStorage<A> for EnvCursor<A> {
+    fn add_eavi(
+        &self,
+        eav: &EntityAttributeValueIndex<A>,
+    ) -> PersistenceResult<Option<EntityAttributeValueIndex<A>>> {
+        self.staging_eav_db.add_lmdb_eavi(eav).map_err(to_api_error)
+    }
+
+    fn fetch_eavi(
+        &self,
+        query: &EaviQuery<A>,
+    ) -> PersistenceResult<BTreeSet<EntityAttributeValueIndex<A>>> {
+        let eavis = self.staging_eav_db.fetch_eavi(query)?;
+
+        if !eavis.is_empty() {
+            return Ok(eavis);
+        }
+
+        let eavis = self.eav_db.fetch_eavi(query)?;
+
+        for eavi in &eavis {
+            self.staging_cas_db.add(eavi)?;
+        }
+        Ok(eavis)
+    }
+}
+
+impl<A: Attribute + DeserializeOwned> Cursor<A> for EnvCursor<A> {}
+
+#[derive(Clone)]
+pub struct LmdbCursorProvider<A: Attribute> {
+    cas_db: LmdbStorage,
+    eav_db: EavLmdbStorage<A>,
+    staging_path: PathBuf,
+    staging_initial_map_bytes: Option<usize>,
+}
+
+impl<A: Attribute + DeserializeOwned> CursorProvider<A> for LmdbCursorProvider<A> {
+    type Cursor = EnvCursor<A>;
+    fn create_cursor(&self) -> PersistenceResult<Self::Cursor> {
+        let staging_cas_db_name = format!("STAGING_CAS_{}", Uuid::new_v4());
+        let staging_eav_db_name = format!("STAGING_EAV_{}", Uuid::new_v4());
+        let db_names = vec![staging_cas_db_name.as_str(), staging_eav_db_name.as_str()];
+
+        let staging_dbs = LmdbInstance::new_all(
+            db_names.as_slice(),
+            self.staging_path.clone(),
+            self.staging_initial_map_bytes,
+        );
+
+        let staging_cas_db = LmdbStorage::wrap(staging_dbs.get(&staging_cas_db_name).unwrap());
+        let staging_eav_db = EavLmdbStorage::wrap(staging_dbs.get(&staging_eav_db_name).unwrap());
+
+        Ok(EnvCursor::new(
+            self.cas_db.clone(),
+            self.eav_db.clone(),
+            staging_cas_db,
+            staging_eav_db,
+        ))
+    }
+}
+
+pub fn new_manager<
+    A: Attribute + DeserializeOwned,
+    EP: AsRef<Path> + Clone,
+    SP: AsRef<Path> + Clone,
+>(
+    env_path: EP,
+    staging_path: SP,
+    initial_map_bytes: Option<usize>,
+    staging_initial_map_bytes: Option<usize>,
+) -> DefaultPersistenceManager<A, LmdbStorage, EavLmdbStorage<A>, LmdbCursorProvider<A>> {
+    let cas_db_name = crate::cas::lmdb::CAS_BUCKET;
+    let eav_db_name = crate::eav::lmdb::EAV_BUCKET;
+    let db_names = vec![cas_db_name, eav_db_name];
```

Comment:
Just use an array here. No need to allocate.
```suggestion
    let db_names = [cas_db_name, eav_db_name];
```